README.md
This is an attempt to build fully static, self-contained LLVM tools and libraries.

## Goal

- static libries
  - libc++ (libc++.a, libc++abi.a, libunwind.a)
  - compiler-rt (clang_rt.crtbegin.o, clang_rt.crtend.o, libclang_rt.\*.a)
  - llvm IR, opt etc libs (libLLVMPasses.a, ...)
  - clang libs (libLLVMLibDriver.a, ...)
  - lld libs (libLLVMLinker.a, ...)
- statically-linked tools, linked with the above static libaries:
  - clang
  - lld
  - llvm-ar
  - llvm-ranlib
  - llvm-objdump

You should be able to build your own compiler suite by:

1. making a main.cc file that calls clang_main and/or lld_main
2. compiling that main.cc file with clang & lld
3. statically linking that main object with libc++, clang, lld and llvm libs
   so that the resulting executable does not require a dynamic linker to run
   (exception: on darwin it has to dynamically link with libSystem.dylib, but that's it)


## Build

TL;DR:

```sh
export LLVMBOX_BUILD_DIR=$HOME/tmp
utils/mktmpfs-build-dir.sh 16384 # limit to 16GB

# build stage1 compiler
bash 010-llvm-source-stage1.sh &&
bash 010-zlib-stage1.sh &&
bash 019-llvm-stage1.sh

# build sysroot (some scripts are no-op on non-linux)
bash 020-sysroot.sh &&
bash 021-linux-headers.sh &&
bash 022-musl-libc.sh

# test stage1 compiler
bash 023-test-llvm-stage1.sh

# build packages for stage2 in sysroot
bash 023-musl-fts.sh &&
bash 023-xc.sh &&
bash 023-zlib.sh &&
bash 023-zstd.sh &&
bash 024-libxml2.sh &&
bash 025-xar.sh

# build stage2 compiler
bash 030-llvm-source-stage2.sh
bash 050-llvm-stage2.sh

# test stage2 compiler
bash 090-test-hello.sh

# test full-featured program linking with llvm libs
./myclang/build.sh
```

Host requirements:

- compiler that can build clang (e.g. gcc)
- ninja or equivalent
- cmake 3
- python3
- bash

Tested host systems:

- Ubuntu 20 x86_64
- macOS 10.15 x86_64


If you have a lot of RAM, it is usually much faster to build in a ramfs:

- Linux: `export LLVMBOX_BUILD_DIR=/dev/shm`
- Linux with tmpfs: `mkdir -p ~/tmp && sudo mount -t tmpfs -o size=16G tmpfs ~/tmp && export LLVMBOX_BUILD_DIR=$HOME/tmp`
- macOS: `utils/macos-tmpfs.sh ~/tmp && export LLVMBOX_BUILD_DIR=$HOME/tmp`


Define your build directory

```
export LLVMBOX_BUILD_DIR=build
```

Execute the first section of build steps:

```
for f in $(echo 01*.sh | sort); do bash $f; done
```

Now, there are a few alternatives:


### Alternative 1: custom build scripts

```
for f in $(echo 02*.sh | sort); do bash $f; done
```

Succeeds on macOS


### Alternative 2: "stage2" build

Alternative 1: "stage2" build. This is supposed to be the "correct" way to build an llvm distribution but after about 20 hours of trying to make it work on either macOS and Ubuntu, I have doubts.

```
bash stage2.sh
```


Linux builds passes stage1 but fails stage2, suffering from high memory usage; linker is OOM killed:

```
clang++: error: unable to execute command: Killed
clang++: error: linker command failed due to signal (use -v to see invocation)
```

macOS builds passes stage1 but fails stage2 with libc++ link errors like these:

```
ld64.lld: error: undefined symbol: std::__2::generic_category()
>>> referenced by /tmp/lto.tmp:(symbol std::__2::make_error_code[abi:v15007](std::__2::errc)+0x10)
```

If the linker gets OOM killed, try some of the following ideas:

- set LLVMBOX_LTO_JOBS=N, e.g. `LLVMBOX_LTO_JOBS=2 bash stage2.sh`
- allocate more swap space (if you use zfs, see `utils/linux-swap-zfs.sh`)
- disable LTO: `BOOTSTRAP_LLVM_ENABLE_LTO` in stage1.cmake, `LLVM_ENABLE_LTO` in stage2.cmake (must do a full complete clean rebuild after changing these)



----

## Experiment: build with bazel

```
cd experiments/bazel
bash build-llvm-bazel.sh
```

Uses bazel to build llvm in a sandbox. Build works on linux and mac, but produces invalid results that link to shared libraries on the build host. There seem to be no way of building a "stage2" with LLVM's bazel workspace (see experiments/bazel/targets.txt)


----

## Test

The ultimate test is to be able to build "myclang", our own "custom clang" that links in all the code needed for clang (which is almost everything).

```sh
LLVM_ROOT=path-to-llvm-installation ./myclang/build.sh
```

For example:

```sh
LLVM_ROOT=$LLVMBOX_BUILD_DIR/llvm-x86_64-macos-none ./myclang/build.sh
```


### Test host compiler

After building the host compiler, you should be able to compile C++ programs:

```
$LLVMBOX_BUILD_DIR/llvm-host/bin/clang++ test/hello.cc -o test/hello_cc
test/hello_cc
```

The final program should be statically linked with libc++, ie. not contain any links to c++ libraries (but may contain links to shared host system libc.)

```
$LLVMBOX_BUILD_DIR/llvm-host/bin/llvm-objdump -p test/hello_cc | grep -E 'NEEDED|\.dylib'
```


### Test utilities

There are also wrapper scripts to help with testing, which invokes clang with the appropriate flags:

```sh
LLVM_ROOT=/dev/shm/llvm-host utils/cc test/hello.c -o test/hello_c
LLVM_ROOT=/dev/shm/llvm-host utils/c++ test/hello.cc -o test/hello_cc
```

Another useful utility is `utils/config` which prints compiler and linker flags. For example, it can be used to configure builds:

```sh
#!/bin/sh
export LLVM_ROOT=$HOME/tmp/llvm-x86_64-macos-none
MY_CXXFLAGS="$(utils/config --cxxflags)"
MY_LDFLAGS="$(utils/config --ldflags-cxx)"
"$LLVM_ROOT/bin/clang++" $MY_CXXFLAGS -c main.cc -o main.o
"$LLVM_ROOT/bin/clang++" $MY_LDFLAGS main.o -o myprogram
```


### Testing musl

Testing musl with host compiler (linux only)

```
export LLVMBOX_MUSL=`realpath $(echo $LLVMBOX_BUILD_DIR/musl-$(uname -m)-linux-*)`
$LLVMBOX_BUILD_DIR/llvm-host/bin/clang \
  -isystem $LLVMBOX_MUSL/include \
  -nostartfiles $LLVMBOX_MUSL/lib/crt1.o \
  -nostdlib -L$LLVMBOX_MUSL/lib -lc \
  test/hello.c -o test/hello_c
test/hello_c
# this should not print anything:
$LLVMBOX_BUILD_DIR/llvm-host/bin/llvm-objdump -p test/hello_c | grep NEEDED
```


This dose _NOT_ work for C++ using the host compiler since the host compiler is likely linked with glibc, not musl. I.e:

```
$LLVMBOX_BUILD_DIR/llvm-host/bin/clang++ \
  -nostdinc++ -isystem $LLVMBOX_BUILD_DIR/llvm-host/include/c++/v1 \
  -I$LLVMBOX_BUILD_DIR/llvm-host/include/c++/v1 \
  -I$(echo $LLVMBOX_BUILD_DIR/llvm-host/include/$(uname -m)-unknown-linux-*)/c++/v1 \
  -L$(echo $LLVMBOX_BUILD_DIR/llvm-host/lib/$(uname -m)-unknown-linux-*) \
  -nostdlib++ -L$LLVMBOX_BUILD_DIR/llvm-host/lib \
  -lc++ \
  \
  -isystem $LLVMBOX_MUSL/include \
  -nostartfiles $LLVMBOX_MUSL/lib/crt1.o \
  -nostdlib -L$LLVMBOX_MUSL/lib -lc \
  test/hello.cc -o test/hello_cc

LLVMBOX_BUILD_DIR/llvm-host/include/c++/v1/__locale:572:13: error: unknown type name 'mask'
    bool is(mask __m, char_type __c) const
```



## Issues


Linux llvm host build _"warning: Using 'NAME' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking"_

    [567/3170] Linking CXX executable bin/llvm-config
    /usr/bin/ld: lib/libLLVMSupport.a(Path.cpp.o): in function `llvm::sys::fs::expandTildeExpr(llvm::SmallVectorImpl<char>&)':
    Path.cpp:(.text._ZN4llvm3sys2fsL15expandTildeExprERNS_15SmallVectorImplIcEE+0x1a6): warning: Using 'getpwnam' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking



## Dev notes

### Navigating cmake

ack is useful for looking around for cmake stuff, e.g.

    ack --type=cmake '\bCOMPILER_RT_USE_' ~/tmp/src/llvm


### Core dumps on Ubuntu

Enable saving of core dumps on ubuntu:

```sh
sudo systemctl enable apport.service
sudo service apport start
mkdir -p ~/.config/apport
cat << END >> ~/.config/apport/settings
[main]
unpackaged=true
END
```

Now, when a process crashes:

```sh
rm -rf /tmp/crash && mkdir /tmp/crash
apport-unpack /var/crash/_path_to_program.1000.crash /tmp/crash
ls -l /tmp/crash
(cd /tmp/crash && tar czf ../some-core-dump.tar.gz .)
```



## Useful resources

- https://llvm.org/docs/HowToCrossCompileLLVM.html#hacks
- https://llvm.org/docs/BuildingADistribution.html
- https://libcxx.llvm.org/BuildingLibcxx.html
- https://libcxx.llvm.org/UsingLibcxx.html#alternate-libcxx
- https://github.com/rust-lang/rust/issues/65051
- https://wiki.musl-libc.org/building-llvm.html
- https://fuchsia.googlesource.com/fuchsia/+/master/docs/development/build/toolchain.md
- https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/development/compilers/llvm/11/llvm/default.nix#L277
- https://git.alpinelinux.org/aports/tree/main/llvm15
- https://git.alpinelinux.org/aports/tree/main/llvm-runtimes
- https://git.alpinelinux.org/aports/tree/main/clang15
- https://github.com/Homebrew/homebrew-core/blob/master/Formula/llvm.rb
- [llvm-dev mailing list: "Building LLVM with LLVM with no dependence on GCC"](https://lists.llvm.org/pipermail/llvm-dev/2019-September/135199.html)
- https://libc.llvm.org/full_host_build.html



xar/NEWS
xar Release Notes
=================

1.6.1 (2012-09-17)

  * Fix issues #75, #76
  * Added NEWS file
  * Stealth support for any available EVP_get_digestbyname hash (see man page)
  * Support digests sha224, sha256, sha384 and sha512 provided the linked
    libcrypto supports them (OpenSSL 0.9.8 and later).  Note that using a
    --toc-cksum other than none, md5 or sha1 (the default) is NOT backwards
    compatible (requires extended binary header support), but using a
    --file-cksum other than none, md5 or sha1 is backwards compatible provided
    the recipient's libcrypto supports it.  Both --toc-cksum and --file-cksum
    must be specified to get them to be different.
  * Do not attempt to extract attribute information to a non-file destination
  * Handle 6 more flags bits including UF_HIDDEN and UF_COMPRESSED
  * Fix xar_get_mode output to still include type when mode is unavailable
    (This also fixes alignment problems in -tv output when mode is missing)
  * Fix archival of FinderCreateTime on 64 bit archs (extract was already okay)
  * Add support for --recompress option to blindly recompress all files added
    to the archive not matched by a --no-compress option when a --compression
    type other than none is in effect.  All prior versions of xar always behave
    as though --recompress has been specified.
  * Recognized already-compressed files will now just be stored as-is in the
    archive without a second compression.  The gzip, bzip2 and xz formats are
    currently recognized.  The --recompress option disables this behavior.
    The xar library itself now provides this support (XAR_OPT_RECOMPRESS).
  * xar.h c++ compatibility added
  * XAR_OPT_TOCCKSUM can now be set after adding signatures without causing
    archive corruption
  * Update liblzma support to work with current liblzma versions
  * Show additional information in --version output if --verbose given first
  * Update man page and --help output to be current
  * Allow --compression=xz whenever --compression=lzma is supported
  * Now builds on cygwin
  * A huge number of warnings were eliminated so compilations should be much
    cleaner on many platforms now
  * Add support for --rfc6713 (see man page) (XAR_OPT_RFC6713FORMAT)
  * Support compiling against an older libcrypto but running against a newer one
  * Make extracted-checksum/archived-checksum use long hash name to match value
    used in toc checksum (backwards compatible)
  * Add missing trailing 'Z' to creation-date in archive (backwards compatible)
  * Help output (--help) now goes to standard output instead of standard error
  * Reduce spew on command line argument error
  * Support --directory (-C) during archive as well as extract
  * Add support for --to-stdout (-O) (XAR_OPT_EXTRACTSTDOUT)
  * Add support for --strip-components (XAR_OPT_STRIPCOMPONENTS)
  * Command aliases added (--create, --extract, --list, --file, --directory,
    --verbose, --one-file-system, -V)
  * The xar header version can now be easily #if tested using XAR_VERSION_NUM
  * The xar library version can now be fetched with XAR_OPT_XARLIBVERSION
  * xar_from_base64 no longer adds extra bytes to the decoded output
  * Library now adds a signature-creation-time element which contains the
    "%.1f" formatted number of seconds since 2001-01-01T00:00:00Z when any
    signatures are present
  * Rudimentary command line signature support (--sign/--replace-sign etc.)
  * configure now defaults compiler options to "-Wall -g -O2" for gcc
  * Update Xcode projects
  * Import into git (adding .gitignore and .gitattributes)
  * Initial release from https://github.com/mackyle/xar project

1.5.3 (2010-04-12)

  * Fix issue #68
  * Ensure binary header and toc checksum types match
  * configure improvements

1.5.2 (2007-12-29)

  * Fix issues #2, #32, #41, #42, #43
  * Allow --prop-exclude ea to exclude darwin/FreeBSD extended attributes
  * Add support for --compression-args
  * Add support for lzma compression (-a or --compression=lzma)
  * Add support for --keep-existing (-k)
  * Add support for bzip2 compression (-j or --compression=bzip2)

1.5.0 (2007-05-16)

  * Initial release from https://code.google.com/p/xar/ project


config.sh
#
# Speed things up by building in a ramfs:
#   Linux:
#     export LLVMBOX_BUILD_DIR=/dev/shm
#   Linux: (alt: tmpfs)
#     mkdir -p ~/tmp && sudo mount -o size=16G -t tmpfs none ~/tmp
#     export LLVMBOX_BUILD_DIR=$HOME/tmp
#   macOS:
#     ./macos-tmpfs.sh ~/tmp
#     export LLVMBOX_BUILD_DIR=$HOME/tmp
#
set -e
_err() { echo "$0:" "$@" >&2; exit 1; }
PWD0=${PWD0:-$PWD}
[ -n "${LLVMBOX_BUILD_DIR:-}" ] || _err "LLVMBOX_BUILD_DIR is not set in env"

PROJECT=${LLVMBOX_PROJECT:-$(realpath "$(dirname "$0")")}
BUILD_DIR=$(realpath "$LLVMBOX_BUILD_DIR")
DOWNLOAD_DIR=$(realpath "${LLVMBOX_DOWNLOAD_DIR:-$PROJECT/download}")
OUT_DIR=$(realpath "${LLVMBOX_OUT_DIR:-$PROJECT/out}")
NCPU=${LLVMBOX_NCPU:-$(nproc)}
HOST_SYS=$(uname -s)
HOST_ARCH=$(uname -m)

# apple still uses the legacy name "arm64" (renamed to aarch64 in llvm, may 2014)
[ "$HOST_ARCH" != "arm64" ] || HOST_ARCH=aarch64

# ————————————————————————————————————————————————————————————————————————————————————

_version_gte() { # <v> <minv>
  local v1 v2 v3 min_v1 min_v2 min_v3
  IFS=. read -r v1 v2 v3 <<< "$1"
  IFS=. read -r min_v1 min_v2 min_v3 <<< "$2"
  [ -n "$min_v2" ] || min_v2=0
  [ -n "$min_v3" ] || min_v3=0
  [ -n "$v2" ] || v2=$min_v2
  [ -n "$v3" ] || v3=$min_v3
  # echo "v   $v1 $v2 $v3"
  # echo "min $min_v1 $min_v2 $min_v3"
  if [ "$v1" -lt "$min_v1" ]; then return 1; fi
  if [ "$v1" -gt "$min_v1" ]; then return 0; fi
  if [ "$v2" -lt "$min_v2" ]; then return 1; fi
  if [ "$v2" -gt "$min_v2" ]; then return 0; fi
  if [ "$v3" -lt "$min_v3" ]; then return 1; fi
  if [ "$v3" -gt "$min_v3" ]; then return 0; fi
}

TARGET=${LLVMBOX_TARGET:-}  # e.g. aarch64-linux, x86_64-macos, x86_64-macos.11.7
TARGET_ARCH=                # e.g. x86_64, aarch64
TARGET_SYS=                 # e.g. linux, macos, macos.11.7
TARGET_SYS_VERSION=         # e.g. 11.7 (from macos.11.7) -- OS ABI (M.m[.p])
TARGET_SYS_MINVERSION=      # compatibility version of OS ABI
TARGET_SYS=                 # e.g. macos (from macos.11.7)
TARGET_TRIPLE=              # e.g. x86_64-linux-musl (for clang/llvm)
TARGET_CMAKE_SYSTEM_NAME=   # e.g. Linux
if [ -z "$TARGET" ]; then
  case "$HOST_SYS" in
    Darwin) TARGET=$HOST_ARCH-macos ;;
    Linux)  TARGET=$HOST_ARCH-linux ;;
    *)      _err "couldn't guess TARGET from $HOST_SYS"
  esac
fi
TARGET_ARCH=${TARGET%-*}
TARGET_SYS=${TARGET#*-}
if [[ "$TARGET_SYS" == *"."* ]]; then
  TARGET_SYS_VERSION=${TARGET_SYS#*.} # macos.11.7 => 11.7
  TARGET_SYS=${TARGET_SYS%%.*}        # macos.11.7 => macos
fi
[ "$TARGET_SYS" = native ] && case "$HOST_SYS" in
  Darwin) TARGET_SYS=macos ;;
  Linux)  TARGET_SYS=linux ;;
  *) _err "$TARGET: couldn't guess TARGET_SYS from HOST_SYS=$HOST_SYS"
esac
case "$TARGET_SYS" in
  macos)
    TARGET_CMAKE_SYSTEM_NAME="Darwin"
    TARGET_TRIPLE=$TARGET_ARCH-apple-darwin ;;
  freebsd)
    TARGET_CMAKE_SYSTEM_NAME="FreeBSD"
    TARGET_TRIPLE=$TARGET_ARCH-freebsd-gnu ;;
  windows)
    TARGET_CMAKE_SYSTEM_NAME="Windows"
    TARGET_TRIPLE=$TARGET_ARCH-windows-gnu ;;
  linux)
    TARGET_CMAKE_SYSTEM_NAME="Linux"
    TARGET_TRIPLE=$TARGET_ARCH-linux-musl ;;
  *) _err "unsupported TARGET system '$TARGET_SYS' in '$TARGET'"
esac
# set TARGET_SYS_MINVERSION (oldest OS ABI we support.)
# Note: macOS SDKs didn't ship with complete libSystem.tbd files until 10.15
case "$TARGET_ARCH-$TARGET_SYS" in
  x86_64-macos)  TARGET_SYS_MINVERSION=10.15 ;;
  aarch64-macos) TARGET_SYS_MINVERSION=11.0 ;;
  *)             TARGET_SYS_MINVERSION=1.0 ;;
esac
# check so that TARGET_SYS_VERSION >= TARGET_SYS_MINVERSION
if [ -z "$TARGET_SYS_VERSION" ]; then
  TARGET_SYS_VERSION=$TARGET_SYS_MINVERSION
elif [[ "$TARGET_SYS_VERSION" =~ ^[0-9]+(\.[0-9]+(\.[0-9]+)?)?$ ]]; then
  _version_gte "$TARGET_SYS_VERSION" "$TARGET_SYS_MINVERSION" ||
    _err "TARGET version $TARGET_SYS_VERSION too old, older than the minimum $TARGET_SYS_MINVERSION"
else
  _err "invalid TARGET version format '$TARGET_SYS_VERSION'; expected M[.m[.p]]"
fi
TARGET_SYS_VERSION_MAJOR=${TARGET_SYS_VERSION%%.*}  # e.g. 1 in 1.2.3
# rewrite TARGET to canonical form arch-sysname-sysversionmajor
TARGET=$TARGET_ARCH-$TARGET_SYS-$TARGET_SYS_VERSION_MAJOR

SYSROOTS_DIR=$PROJECT/sysroots
LLVMBOX_SYSROOT_BASE=${LLVMBOX_SYSROOT_BASE:-$OUT_DIR/sysroot}
LLVMBOX_SYSROOT=${LLVMBOX_SYSROOT:-$LLVMBOX_SYSROOT_BASE/$TARGET}
export LLVMBOX_SYSROOT

# ————————————————————————————————————————————————————————————————————————————————————

LLVM_RELEASE=15.0.7
LLVM_SHA256=42a0088f148edcf6c770dfc780a7273014a9a89b66f357c761b4ca7c8dfa10ba
LLVM_SRC_URL=https://github.com/llvm/llvm-project/archive/llvmorg-${LLVM_RELEASE}.tar.gz
[[ "$LLVM_RELEASE" != *"."* ]] && # git snapshot
  LLVM_SRC_URL=https://github.com/llvm/llvm-project/archive/${LLVM_RELEASE}.tar.gz
LLVM_SRC_STAGE1=${LLVM_SRC_STAGE1:-$OUT_DIR/src/llvm-stage1}
LLVM_SRC=${LLVM_SRC:-$OUT_DIR/src/llvm}
LLVM_STAGE1=${LLVM_STAGE1:-$OUT_DIR/llvm-stage1}

ZLIB_VERSION=1.2.13
ZLIB_SHA256=b3a24de97a8fdbc835b9833169501030b8977031bcb54b3b3ac13740f846ab30
ZLIB_SRC=${ZLIB_SRC:-$BUILD_DIR/src/zlib}
ZLIB_STAGE1=${ZLIB_HOST:-$BUILD_DIR/zlib-stage1}
ZLIB_DIST=${ZLIB_DIST:-$BUILD_DIR/zlib-$TARGET}

ZSTD_VERSION=1.5.2
ZSTD_SHA256=7c42d56fac126929a6a85dbc73ff1db2411d04f104fae9bdea51305663a83fd0
ZSTD_SRC=${ZSTD_SRC:-$BUILD_DIR/src/zstd}
ZSTD_DIST=${ZSTD_DIST:-$BUILD_DIR/zstd-$TARGET}

XC_VERSION=5.2.5
XC_SHA256=3e1e518ffc912f86608a8cb35e4bd41ad1aec210df2a47aaa1f95e7f5576ef56
XC_SRC=${XC_SRC:-$BUILD_DIR/src/xc}
XC_DESTDIR=${XC_DESTDIR:-$BUILD_DIR/xc-$TARGET}

OPENSSL_VERSION=1.1.1s
OPENSSL_SHA256=c5ac01e760ee6ff0dab61d6b2bbd30146724d063eb322180c6f18a6f74e4b6aa
OPENSSL_SRC=${OPENSSL_SRC:-$BUILD_DIR/src/openssl}
OPENSSL_DESTDIR=${OPENSSL_DESTDIR:-$BUILD_DIR/openssl-$TARGET}

LIBXML2_VERSION=2.10.3
LIBXML2_SHA256=5d2cc3d78bec3dbe212a9d7fa629ada25a7da928af432c93060ff5c17ee28a9c
LIBXML2_SRC=${LIBXML2_SRC:-$BUILD_DIR/src/libxml2}
LIBXML2_DESTDIR=${LIBXML2_DESTDIR:-$BUILD_DIR/libxml2-$TARGET}

XAR_SRC=${XAR_SRC:-$BUILD_DIR/src/xar}
XAR_DESTDIR=${XAR_DESTDIR:-$BUILD_DIR/xar-$TARGET}

LINUX_VERSION=6.1.7
LINUX_VERSION_MAJOR=${LINUX_VERSION%%.*}
LINUX_SHA256=4ab048bad2e7380d3b827f1fad5ad2d2fc4f2e80e1c604d85d1f8781debe600f
LINUX_SRC=${LINUX_SRC:-$BUILD_DIR/src/linux}
LINUX_HEADERS_DESTDIR=${LINUX_HEADERS_DESTDIR:-$BUILD_DIR/linux-${LINUX_VERSION}-headers}

MUSLFTS_SRC=${MUSLFTS_SRC:-$BUILD_DIR/src/musl-fts}
MUSLFTS_DESTDIR=${MUSLFTS_DESTDIR:-$BUILD_DIR/musl-fts-$TARGET}

MUSL_VERSION=1.2.3
MUSL_SHA256=7d5b0b6062521e4627e099e4c9dc8248d32a30285e959b7eecaa780cf8cfd4a4
MUSL_SRC=${MUSL_SRC:-$BUILD_DIR/src/musl}
MUSL_HOST=${MUSL_HOST:-$BUILD_DIR/musl-host}
MUSL_DESTDIR=${MUSL_DESTDIR:-$BUILD_DIR/musl-$TARGET}

GCC_MUSL=${GCC_MUSL:-$BUILD_DIR/gcc-musl}

# ————————————————————————————————————————————————————————————————————————————————————

STAGE1_CC=cc
STAGE1_CXX=c++
STAGE1_AR=ar
STAGE1_RANLIB=ranlib
STAGE1_CFLAGS=
STAGE1_LDFLAGS=
if [ "$HOST_SYS" = "Linux" ]; then
  STAGE1_CC="$(command -v  gcc || true)"
  STAGE1_CXX="$(command -v g++ || true)"
  STAGE1_LDFLAGS="-static-libgcc -static"
  # STAGE1_CC="$OUT_DIR/gcc-musl/bin/gcc"
  # STAGE1_CXX="$OUT_DIR/gcc-musl/bin/g++"
  # STAGE1_AR="$OUT_DIR/gcc-musl/bin/ar"
  # STAGE1_RANLIB="$OUT_DIR/gcc-musl/bin/ranlib"
  # STAGE1_LD="$OUT_DIR/gcc-musl/bin/ld"
  # STAGE1_LDFLAGS="-static-libgcc -static"
elif [ "$HOST_SYS" = "Darwin" ]; then
  HOST_MACOS_SDK=$(xcrun -sdk macosx --show-sdk-path)
  [ -d "$HOST_MACOS_SDK" ] || _err "macOS SDK not found. Try: xcode-select --install"
  STAGE1_CC="$(command -v  clang || true)"
  STAGE1_CXX="$(command -v clang++ || true)"
  STAGE1_MACOS_VERSION=10.15
  [ "$HOST_ARCH" != x86_64 ] && STAGE1_MACOS_VERSION=11.7
  STAGE1_CFLAGS="$STAGE1_CFLAGS -mmacosx-version-min=$TARGET_SYS_MINVERSION"
  STAGE1_LDFLAGS="$STAGE1_LDFLAGS -mmacosx-version-min=$TARGET_SYS_MINVERSION"
fi
STAGE1_ASM=${STAGE1_ASM:-$STAGE1_CC}
STAGE1_LD=${STAGE1_LD:-$STAGE1_CXX}
# canonicalize paths and check that all tools exist
for tool in CC CXX AR RANLIB LD; do
  var=STAGE1_$tool
  tool=${!var}
  [[ "$tool" == *"/"* ]] || declare "$var=$(command -v "$tool" || true)"
  tool=${!var}
  [ -x "$tool" ] || _err "$tool not found ($var)"
done

# ————————————————————————————————————————————————————————————————————————————————————

STAGE2_CC="$LLVM_STAGE1/bin/clang"
STAGE2_CXX="$LLVM_STAGE1/bin/clang++"
STAGE2_ASM=$STAGE2_CC
STAGE2_LD=$STAGE2_CC
STAGE2_RC="$LLVM_STAGE1/bin/llvm-rc"
STAGE2_AR="$LLVM_STAGE1/bin/llvm-ar"
STAGE2_RANLIB="$LLVM_STAGE1/bin/llvm-ranlib"

STAGE2_CFLAGS=(
  --target=$TARGET_TRIPLE \
  --sysroot="$LLVMBOX_SYSROOT" \
  -isystem "$LLVMBOX_SYSROOT/include" \
)
STAGE2_LDFLAGS=(
  --target=$TARGET_TRIPLE \
  --rtlib=compiler-rt \
)
case "$TARGET_SYS" in
  macos)
    STAGE2_CFLAGS+=( -mmacosx-version-min=$TARGET_SYS_VERSION )
    STAGE2_LDFLAGS+=( -mmacosx-version-min=$TARGET_SYS_VERSION )
    STAGE2_LDFLAGS+=( -L"$HOST_MACOS_SDK/usr/lib" )
    # STAGE2_CFLAGS+=( -mmacosx-version-min=$TARGET_SYS_MINVERSION \
    #   -isystem "$HOST_MACOS_SDK/usr/include" )
    # STAGE2_LDFLAGS+=( -mmacosx-version-min=$TARGET_SYS_MINVERSION )
    # STAGE2_CFLAGS+=( -I"$LLVMBOX_SYSROOT/include" )
    # STAGE2_LDFLAGS+=( -L"$LLVMBOX_SYSROOT/lib" )
    ;;
  *)
    STAGE2_LDFLAGS+=( --sysroot="$LLVMBOX_SYSROOT" )
esac

# ————————————————————————————————————————————————————————————————————————————————————
# functions

_relpath() { # <path>
  case "$1" in
    "$PWD0/"*) echo "${1##${2:-$PWD0}/}" ;;
    "$PWD0")   echo "." ;;
    *)         echo "${1/$HOME\//~/}" ;;
  esac
}

_pushd() {
  pushd "$1" >/dev/null
  [ "$PWD" = "$PWD0" ] || echo "cd $(_relpath "$PWD")"
}

_popd() {
  popd >/dev/null
  [ "$PWD" = "$PWD0" ] || echo "cd $(_relpath "$PWD")"
}

_array_join() { # <gluechar> <element> ...
  local IFS="$1"; shift; echo "$*"
}

_sha_test() { # <file> [<sha256> | <sha512>]
  local file=$1 ; local expect=$2
  [ -f "$file" ] || return 1
  case "${#expect}" in
    128) kind=512; actual=$(sha512sum "$file" | cut -d' ' -f1) ;;
    64)  kind=256; actual=$(sha256sum "$file" | cut -d' ' -f1) ;;
    *)   _err "checksum $expect has incorrect length (not sha256 nor sha512)" ;;
  esac
  [ "$actual" = "$actual" ] || return 1
}

_sha_verify() { # <file> [<sha256> | <sha512>]
  local file=$1
  local expect=$2
  local actual=
  case "${#expect}" in
    128) kind=512; actual=$(sha512sum "$file" | cut -d' ' -f1) ;;
    64)  kind=256; actual=$(sha256sum "$file" | cut -d' ' -f1) ;;
    *)   _err "checksum $expect has incorrect length (not sha256 nor sha512)" ;;
  esac
  if [ "$actual" != "$expect" ]; then
    echo "$file: SHA-$kind sum mismatch:" >&2
    echo "  actual:   $actual" >&2
    echo "  expected: $expect" >&2
    return 1
  fi
}

_download_nocache() { # <url> <outfile> [<sha256> | <sha512>]
  local url=$1 ; local outfile=$2 ; local checksum=${3:-}
  rm -f "$outfile"
  mkdir -p "$(dirname "$outfile")"
  echo "${outfile##$PWD0/}: fetch $url"
  command -v wget >/dev/null &&
    wget -q --show-progress -O "$outfile" "$url" ||
    curl -L '-#' -o "$outfile" "$url"
  [ -z "$checksum" ] || _sha_verify "$outfile" "$checksum"
}

_download() { # <url> <outfile> [<sha256> | <sha512>]
  local url=$1 ; local outfile=$2 ; local checksum=${3:-}
  if [ -f "$outfile" -a -z "$checksum" ] || _sha_test "$outfile" "$checksum"; then
    return 0
  fi
  _download_nocache "$url" "$outfile" "$checksum"
}

_extract_tar() { # <file> <outdir>
  [ $# -eq 2 ] || _err "_extract_tar"
  local tarfile=$1
  local outdir=$2
  [ -e "$tarfile" ] || _err "$tarfile not found"

  local extract_dir="${outdir%/}-extract-$(basename "$tarfile")"
  rm -rf "$extract_dir"
  mkdir -p "$extract_dir"

  echo "${outdir##$PWD0/}: extract ${tarfile##$PWD0/}"
  if ! XZ_OPT='-T0' tar -C "$extract_dir" -xf "$tarfile"; then
    rm -rf "$extract_dir"
    return 1
  fi
  rm -rf "$outdir"
  mkdir -p "$(dirname "$outdir")"
  mv -f "$extract_dir"/* "$outdir"
  rm -rf "$extract_dir"
}

_fetch_source_tar() { # <url> (<sha256> | <sha512> | "") <outdir> [<tarfile>]
  [ $# -gt 2 ] || _err "_fetch_source_tar ($#)"
  local url=$1
  local checksum=$2
  local outdir=$3
  local tarfile=${4:-$DOWNLOAD_DIR/$(basename "$url")}
  _download    "$url" "$tarfile" "$checksum"
  _extract_tar "$tarfile" "$outdir"
}
